providers/attach.go

```diff
@@ -0,0 +1,48 @@
+package providers
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+)
+
+func StartProviders(ctx context.Context, factories map[string]ProviderFactory) (map[string]int, context.CancelFunc, error) {
```

Comment:
What is the string in factories? A bit of documentation can go a long way. AS well as what is `map[string]int`? Is this like "aws"? aNd ports?

Reply:
Yup ... adding better typing here 👍